File: leetcode/Gas_Station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost) {
        vector<int> left(gas.size());
        
        for(int i = 0; i < gas.size(); i++){
            left[i] = gas[i] - cost[i];
        }
        
        int s = 0;
        int sum = 0;
        int t_sum = 0;
        
        for(int i = 0; i < left.size(); i++){
            sum += left[i];
            t_sum += left[i];
            
            if(sum < 0){
                sum = 0;
                s = i+1;
            }
        }
        if(s >= left.size() || t_sum < 0) return -1;
        else return s;
        
    }
};

File: leetcode/java/GasStation.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int sum = 0;
        int n = gas.length;
        int pSum = 0;
        int cand = -1;
        
        for(int i = 0; i < n; i++) {
            int t = gas[i] - cost[i];
            sum += t;
            pSum += t;
            
            if(pSum < 0) {
                cand = -1;
                pSum = 0;
            }else{
                if(cand == -1) cand = i;
            }
        }
        if (sum < 0) return -1;
        return cand;
    }
}
